File: src/main/java/li/klass/fhem/adapter/uiservice/StateUiService.java
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2011, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLIC LICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301  USA
 */

package li.klass.fhem.adapter.uiservice;

import android.content.Context;
import android.content.Intent;

import javax.inject.Inject;
import javax.inject.Singleton;

import li.klass.fhem.adapter.devices.core.UpdatingResultReceiver;
import li.klass.fhem.constants.Actions;
import li.klass.fhem.constants.BundleExtraKeys;
import li.klass.fhem.domain.core.DimmableDevice;
import li.klass.fhem.domain.core.FhemDevice;
import li.klass.fhem.service.intent.DeviceIntentService;

@Singleton
public class StateUiService {
    @Inject
    public StateUiService() {
    }


    public void setSubState(FhemDevice<?> device,
                            String stateName, String value, Context context) {
        context.startService(new Intent(Actions.DEVICE_SET_SUB_STATE)
                .setClass(context, DeviceIntentService.class)
                .putExtra(BundleExtraKeys.DEVICE_NAME, device.getName())
                .putExtra(BundleExtraKeys.STATE_NAME, stateName)
                .putExtra(BundleExtraKeys.STATE_VALUE, value)
                .putExtra(BundleExtraKeys.RESULT_RECEIVER, new UpdatingResultReceiver(context)));
    }

    public void setState(FhemDevice<?> device, String value, Context context) {
        context.startService(new Intent(Actions.DEVICE_SET_STATE)
                .setClass(context, DeviceIntentService.class)
                .putExtra(BundleExtraKeys.DEVICE_TARGET_STATE, value)
                .putExtra(BundleExtraKeys.DEVICE_NAME, device.getName())
                .putExtra(BundleExtraKeys.RESULT_RECEIVER, new UpdatingResultReceiver(context)));
    }

    public void setDim(DimmableDevice<?> device, float progress, Context context) {
        context.startService(new Intent(Actions.DEVICE_DIM)
                .setClass(context, DeviceIntentService.class)
                .putExtra(BundleExtraKeys.DEVICE_DIM_PROGRESS, progress)
                .putExtra(BundleExtraKeys.DEVICE_NAME, device.getName())
                .putExtra(BundleExtraKeys.RESULT_RECEIVER, new UpdatingResultReceiver(context)));
    }
}

File: app/src/main/java/li/klass/fhem/adapter/uiservice/StateUiService.kt
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2011, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLIC LICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301  USA
 */

package li.klass.fhem.adapter.uiservice

import android.content.Context
import android.content.Intent
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import li.klass.fhem.constants.Actions
import li.klass.fhem.devices.backend.GenericDeviceService
import li.klass.fhem.domain.core.FhemDevice
import li.klass.fhem.update.backend.xmllist.XmlListDevice
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class StateUiService @Inject constructor(
        val genericDeviceService: GenericDeviceService
) {

    suspend fun setSubState(device: XmlListDevice,
                            stateName: String, value: String, connectionId: String?, context: Context) {
        coroutineScope {
            withContext(Dispatchers.IO) {
                genericDeviceService.setSubState(device, stateName, value, connectionId)
            }
            invokeUpdate(context)
        }
    }

    suspend fun setState(device: FhemDevice, value: String, context: Context, connectionId: String?) =
            setState(device.xmlListDevice, value, context, connectionId)

    suspend fun setState(device: XmlListDevice, value: String, context: Context, connectionId: String?) {
        coroutineScope {
            withContext(Dispatchers.IO) {
                genericDeviceService.setState(device, value, connectionId)
            }
            invokeUpdate(context)
        }
    }

    private fun invokeUpdate(context: Context) = context.sendBroadcast(Intent(Actions.DO_UPDATE).apply { setPackage(context.packageName) })
}
